File: cpp/src/image_io.cpp
#include "pdi/image_io.hpp"
#include "pdi/errors.hpp"
#include "pdi/result_io.hpp"

#include <opencv2/imgcodecs.hpp>
#include <stdexcept>

namespace pdi {

cv::Mat read_image(const std::string& path)
{
    // IMREAD_UNCHANGED preserva o número de canais e a profundidade do arquivo.
    // Isso é importante no M1.1, pois o estudante precisa observar como a
    // representação da imagem influencia o acesso aos pixels.
    cv::Mat image = cv::imread(path, cv::IMREAD_UNCHANGED);

    if (image.empty()) {
        throw PdiError(ExitCode::read_error, "Nao foi possivel abrir a imagem: " + path);
    }

    return image;
}

void write_image(const std::string& path, const cv::Mat& image)
{
    if (image.empty()) {
        throw PdiError(ExitCode::write_error, "A imagem de saida esta vazia.");
    }

    try {
        // A infraestrutura cria o diretório-pai quando necessário para que o
        // estudante possa concentrar-se no algoritmo e não em detalhes de I/O.
        ensure_parent_directory(path);
    } catch (const std::exception& error) {
        throw PdiError(
            ExitCode::write_error,
            "Nao foi possivel preparar o caminho de saida: " + std::string(error.what())
        );
    }

    if (!cv::imwrite(path, image)) {
        throw PdiError(ExitCode::write_error, "Nao foi possivel salvar a imagem: " + path);
    }
}

} // namespace pdi

File: python/src/pdi_lab/image_io.py
from __future__ import annotations

import cv2
import numpy as np

from . import exit_code
from .errors import PdiError
from .result_io import ensure_parent_directory


def read_image(path: str) -> np.ndarray:
    """Lê a imagem preservando canais e profundidade do arquivo."""

    image = cv2.imread(path, cv2.IMREAD_UNCHANGED)
    if image is None:
        raise PdiError(exit_code.READ_ERROR, f"Nao foi possivel abrir a imagem: {path}")
    return image


def write_image(path: str, image: np.ndarray) -> None:
    """Salva a imagem e cria automaticamente os diretórios-pai."""

    if image is None or image.size == 0:
        raise PdiError(exit_code.WRITE_ERROR, "A imagem de saida esta vazia.")

    try:
        output = ensure_parent_directory(path)
    except OSError as error:
        raise PdiError(
            exit_code.WRITE_ERROR,
            f"Nao foi possivel preparar o caminho de saida: {error}",
        ) from error

    if not cv2.imwrite(str(output), image):
        raise PdiError(exit_code.WRITE_ERROR, f"Nao foi possivel salvar a imagem: {path}")
